pkg/rebalancer/storage/controller/service/rebalancer.go

```diff
@@ -19,28 +19,526 @@ package service
+	"github.com/vdaas/vald/internal/k8s"
+	"github.com/vdaas/vald/internal/k8s/configmap"
+	"github.com/vdaas/vald/internal/k8s/decoder"
+	"github.com/vdaas/vald/internal/k8s/job"
+	mpod "github.com/vdaas/vald/internal/k8s/metrics/pod"
+	"github.com/vdaas/vald/internal/k8s/pod"
+	"github.com/vdaas/vald/internal/k8s/statefulset"
+	"github.com/vdaas/vald/internal/log"
+	"github.com/vdaas/vald/internal/safety"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/config"
+	"github.com/vdaas/vald/pkg/rebalancer/storage/controller/model"
 )
 
+const (
+	qualifiedNamePrefix string = "rebalancer.vald.vdaas.org/"
+)
+
+// Rebalancer represents the rebalancer interface.
 type Rebalancer interface {
-	Start(context.Context) (<-chan error, error)
+	Start(ctx context.Context) (<-chan error, error)
 }
 
-type rebalancer struct{}
+type rebalancer struct {
+	podName      string
+	podNamespace string
+
+	jobs           atomic.Value
+	jobName        string
+	jobNamespace   string
+	jobTemplate    atomic.Value
+	jobTemplateKey string // config map key of job template
+
+	agentName         string
+	agentNamespace    string
+	agentResourceType config.AgentResourceType
+	pods              atomic.Value
+	podMetrics        atomic.Value
+
+	jobConfigmapName string
+
+	statefulSets atomic.Value
+
+	leaderElectionID string
 
-func New(opts ...Option) (dsc Rebalancer, err error) {
+	rcd           time.Duration // reconcile check duration
+	eg            errgroup.Group
+	ctrl          k8s.Controller
+	tolerance     float64
+	rateThreshold float64
+	decoder       *decoder.Decoder
+}
+
+// NewRebalancer initialize job, configmap, pod, podMetrics, statefulset, replicaset, daemonset reconciler.
+// And it returns the rebalancer implemenation or any error occurred.
+func NewRebalancer(opts ...RebalancerOption) (Rebalancer, error) {
 	r := new(rebalancer)
-	for _, opt := range append(defaultOpts, opts...) {
+
+	for _, opt := range append(defaultRebalancerOpts, opts...) {
 		if err := opt(r); err != nil {
 			return nil, errors.ErrOptionFailed(err, reflect.ValueOf(opt))
 		}
 	}
+
+	job, err := job.New(
+		job.WithControllerName("job rebalancer"),
+		job.WithNamespaces(r.jobNamespace),
+		job.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		job.WithOnReconcileFunc(func(jobList map[string][]job.Job) {
+			log.Debugf("[reconcile Job] length Joblist: %d", len(jobList))
+			jobs, ok := jobList[r.jobName]
+			if ok {
+				r.jobs.Store(jobs)
+			} else {
+				r.jobs.Store(make([]job.Job, 0))
+				log.Infof("job not found: %s", r.jobName)
+			}
+		}),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cm, err := configmap.New(
+		configmap.WithControllerName("configmap rebalancer"),
+		configmap.WithNamespaces(r.jobNamespace),
+		configmap.WithOnErrorFunc(func(err error) {
+			log.Error(err)
+		}),
+		configmap.WithOnReconcileFunc(func(configmapList map[string][]configmap.ConfigMap) {
+			configmaps, ok := configmapList[r.jobNamespace]
+			if ok {
```

Comment:
It is better to use early return instead of if/else.

Reply:
thank you
I fixed: https://github.com/vdaas/vald/pull/1187/commits/d42716a5312d36707b80cde6a1992220b17b8ed4